src/Symfony/Component/HttpClient/RetryHttpClient.php

```diff
@@ -0,0 +1,94 @@
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\Chunk\FirstChunk;
+use Symfony\Component\HttpClient\Response\AsyncContext;
+use Symfony\Component\HttpClient\Response\AsyncResponse;
+use Symfony\Component\HttpClient\Response\MockResponse;
+use Symfony\Component\HttpClient\Retry\RetryStrategyInterface;
+use Symfony\Contracts\HttpClient\ChunkInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Automatically retries the failing requests.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RetryHttpClient implements HttpClientInterface
+{
+    use AsyncDecoratorTrait;
+
+    private $strategy;
+    private $logger;
+
+    private $retryCount = 0;
+    private $pausing = false;
+
+    public function __construct(HttpClientInterface $client, RetryStrategyInterface $strategy, LoggerInterface $logger = null)
+    {
+        $this->client = $client;
+        $this->strategy = $strategy;
+        $this->logger = $logger ?: new NullLogger();
+    }
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options) {
+            // resume from pause
+            if ($this->pausing) {
+                $this->pausing = false;
+
+                $context->replaceRequest($method, $url, $options);
+
+                return;
+            }
+            $exception = null;
+            try {
+                // only retry first chunk
+                if ($chunk->isTimeout() || !$chunk->isFirst()) {
+                    yield $chunk;
+
+                    return;
+                }
+            } catch (TransportExceptionInterface $exception) {
+                // catch TransportExceptionInterface to send it to strategy.
+            }
+
+            if (!$this->strategy->isRetryable($this->retryCount, $response = $context->getResponse(), $exception)) {
+                yield $chunk;
+
+                return;
+            }
+
+            ++$this->retryCount;
+
+            $context->setInfo('retry', $this->retryCount);
+            $response->cancel();
+
+            $delay = $this->strategy->getWaitingTime($this->retryCount, $response, $exception);
+
+            $this->logger->info('Error returned by the server. Retrying #{retryCount} using {delay} ms delay. Error: "{error}"', [
+                'retryCount' => $this->retryCount,
+                'delay' => $delay,
+                'error' => ($exception ? $exception->getMessage() : 'StatusCode: '.$response->getStatusCode()),
+            ]);
+
+            $this->pausing = true;
+            $context->pause($delay / 1000);
```

Comment:
I tried to use `pause` but did not achieved to make it worked:

- when calling `$response->cancel();` at line 80, I get a `A chunk passthru must yield an "isLast()" chunk before ending a stream.` error
- when removing the `cancel` call, There is no pause, it calls immediatly the callback with a new chunk

any pointer?

Reproduct case:
```
        $client = new RetryHttpClient(HttpClient::create(), new MultiplierRetryStrategy(), new ConsoleLogger(new ConsoleOutput(256)));
        $client->request('GET', 'https://httpstat.us/500')->getContent();
```

Reply:
you need to 1. replace the request 2. pause it

Reply:
it works. Thanks